device.go

```diff
@@ -161,6 +161,142 @@ func (device *VFIODevice) deviceType() string {
 	return device.DeviceType
 }
 
+// VhostUserDeviceType - represents a vhost-user device type
+// Currently support just VhostUserNet
+type VhostUserDeviceType string
+
+const (
+	//VhostUserSCSI - SCSI based vhost-user type
+	VhostUserSCSI = "vhost-user-scsi-pci"
+	//VhostUserNet - net based vhost-user type
+	VhostUserNet = "virtio-net-pci"
+	//VhostUserBlk represents a block vhostuser device type
+	VhostUserBlk = "vhost-user-blk-pci"
+)
+
+// VhostUserDevice represents a vhost-user device. Shared
+// attributes of a vhost-user device can be retrieved using
+// the Attrs() method. Unique data can be obtained by casting
+// the object to the proper type.
+type VhostUserDevice interface {
+	Attrs() *VhostUserDeviceAttrs
+	Type() string
+}
+
+// VhostUserDeviceAttrs represents data shared by most vhost-user devices
+type VhostUserDeviceAttrs struct {
+	DeviceType string
+	DeviceInfo DeviceInfo
+	SocketPath string
+	ID         string
+}
+
+// VhostUserNetDevice is a network vhost-user based device
+type VhostUserNetDevice struct {
+	VhostUserDeviceAttrs
+	MacAddress string
+}
+
+// Attrs returns the VhostUserDeviceAttrs associated with the vhost-user device
+func (vhostUserNetDevice *VhostUserNetDevice) Attrs() *VhostUserDeviceAttrs {
+	return &vhostUserNetDevice.VhostUserDeviceAttrs
+}
+
+// Type returns the type associated with the vhost-user device
+func (vhostUserNetDevice *VhostUserNetDevice) Type() string {
+	return VhostUserNet
+}
+
+// VhostUserSCSIDevice is a SCSI vhost-user based device
+type VhostUserSCSIDevice struct {
+	VhostUserDeviceAttrs
+}
+
+// Attrs returns the VhostUserDeviceAttrs associated with the vhost-user device
+func (vhostUserSCSIDevice *VhostUserSCSIDevice) Attrs() *VhostUserDeviceAttrs {
+	return &vhostUserSCSIDevice.VhostUserDeviceAttrs
+}
+
+// Type returns the type associated with the vhost-user device
+func (vhostUserSCSIDevice *VhostUserSCSIDevice) Type() string {
+	return VhostUserSCSI
+}
+
+// VhostUserBlkDevice is a block vhost-user based device
+type VhostUserBlkDevice struct {
+	VhostUserDeviceAttrs
+}
+
+// Attrs returns the VhostUserDeviceAttrs associated with the vhost-user device
+func (vhostUserBlkDevice *VhostUserBlkDevice) Attrs() *VhostUserDeviceAttrs {
+	return &vhostUserBlkDevice.VhostUserDeviceAttrs
+}
+
+// Type returns the type associated with the vhost-user device
+func (vhostUserBlkDevice *VhostUserBlkDevice) Type() string {
+	return VhostUserBlk
+}
+
+// vhostUserAttach handles the common logic among all of the vhost-user device's
+// attach functions
+func vhostUserAttach(device VhostUserDevice, h hypervisor, c *Container) (err error) {
+	// generate a unique ID to be used for hypervisor commandline fields
+	randBytes, err := generateRandomBytes(8)
+	if err != nil {
+		return err
+	}
+	id := hex.EncodeToString(randBytes)
```

Comment:
does this need the prefix of the type? I see prefixes being used in some places

Reply:
@mcastelino 

No -- the "prefix-ID" is created further down the line.  In our case it is handled in the qemu appendVhostUserDevice function.